File: backend/pages/growth.py
"""The growth page — the chart and the report card.

Both endpoints are thin: the shapes they return are built in
backend/tracking/growth.py, which is where the grading rules live.
"""
from flask import Blueprint, jsonify, render_template, request

from backend.tracking import analytics as analytics_tracking
from backend.tracking import growth as growth_tracking
from backend.tracking import xp as xp_tracking

bp = Blueprint('growth', __name__)


@bp.route('/growth')
def page():
    return render_template('growth.html')


@bp.route('/api/get_growth_data', methods=['GET'])
def get_growth_data():
    """Day-by-day XP, tasks and focus since the account was created."""
    username = request.args.get('username')
    if not username:
        return jsonify({"success": False, "message": "Username required"})

    data = growth_tracking.series(username)
    if data is None:
        return jsonify({"success": False, "message": "User not found"})
    return jsonify({"success": True, **data})


@bp.route('/api/get_growth_ratings', methods=['GET'])
def get_growth_ratings():
    """The five-metric graded report card.

    Computed in tracking/analytics.py, which also files the result away in
    analytics.sql so the grades build up a history.
    """
    username = request.args.get('username')
    if not username:
        return jsonify({"success": False, "message": "Username required"})

    ratings = analytics_tracking.ratings(username)
    if ratings is None:
        return jsonify({"success": False, "message": "User not found"})
    return jsonify({"success": True, **ratings})


@bp.route('/api/get_xp_data', methods=['GET'])
def get_xp_data():
    """The XP ledger rolled up: level, lifetime totals and a per-day series."""
    username = request.args.get('username')
    if not username:
        return jsonify({"success": False, "message": "Username required"})
    return jsonify(xp_tracking.snapshot(username))

File: frontend/src/services/growth.ts
/**
 * Growth: the day-by-day series, the report card, and the XP ledger rolled up.
 *
 * Not in the original services list because the growth page had no service of
 * its own — it called the endpoints directly. It has one now for the same
 * reason the others do: the shapes are fiddly and worth naming once.
 *
 * `ratings()` is not a pure read. Asking for the report card files a dated
 * snapshot per metric into the database, which is how the grades build up a
 * history — so it is worth calling when a page opens and not on a timer.
 *
 * Backend: backend/api/growth.py, backend/tracking/analytics.py.
 */
import { get } from './api';
import type { ApiResult, GrowthDay, Ratings } from '@/types';

export interface GrowthSeries {
  created_date: string;
  days_since_creation: number;
  /** One row per day, including days with nothing on them. */
  growth_data: GrowthDay[];
}

/**
 * The day-by-day series.
 *
 * `days` is how many of the most recent to ask for, and **0 means all of
 * them**. The growth page asks for all and slices client-side: it lets the
 * reader choose 7, 30, 90 or the whole account, and every figure on it that
 * says "vs the previous 30 days" needs the 30 before the 30 on screen. One
 * request that answers every range beats a request per range, and the rows are
 * small.
 */
export function series(
  days = 30,
): Promise<ApiResult<GrowthSeries>> {
  return get<GrowthSeries>('/api/get_growth_data', { days });
}

/** The five-metric graded report card. Files a snapshot as a side effect. */
export function ratings(): Promise<ApiResult<Ratings>> {
  return get<Ratings>('/api/get_growth_ratings');
}

export interface XpSnapshot {
  user: {
    username: string;
    created_date: string;
    days_active: number;
  };
  stats: {
    level: number;
    /** XP within the current level. */
    current_xp: number;
    total_xp: number;
    xp_required: number;
    tasks_completed: number;
  };
  /** Every day since the account was created, not just the last 30. */
  growth_data: Array<
    Pick<
      GrowthDay,
      'date' | 'day_number' | 'xp_earned' | 'tasks_completed' | 'cumulative_xp' | 'avg_task_xp'
    >
  >;
  summary: {
    total_days: number;
    total_xp: number;
    average_xp_per_day: number;
    most_productive_day: GrowthDay | null;
  };
}

export function xpSnapshot(): Promise<ApiResult<XpSnapshot>> {
  return get<XpSnapshot>('/api/get_xp_data');
}
